File: frontend/lib/models/hospital.dart
/// A hospital / clinic that doctors are affiliated with.
class Hospital {
  final String id;
  final String name;
  final String address;
  final String city;
  final double rating;
  final String phone;
  final String imageUrl;
  final List<String> galleryUrls;
  final List<String> departments;
  final List<HospitalFacility> facilities;
  final String about;
  final String openHours; // e.g. "Open 24 hours"
  final double distanceKm;
  final double latitude;
  final double longitude;

  const Hospital({
    required this.id,
    required this.name,
    required this.address,
    required this.city,
    required this.rating,
    required this.phone,
    required this.imageUrl,
    required this.galleryUrls,
    required this.departments,
    required this.facilities,
    required this.about,
    required this.openHours,
    required this.distanceKm,
    required this.latitude,
    required this.longitude,
  });

  factory Hospital.fromJson(Map<String, dynamic> json) => Hospital(
        id: (json['id'] ?? json['_id'] ?? '') as String,
        name: (json['name'] ?? '') as String,
        address: (json['address'] ?? '') as String,
        city: (json['city'] ?? '') as String,
        rating: ((json['rating'] ?? 0) as num).toDouble(),
        phone: (json['phone'] ?? '') as String,
        imageUrl: (json['imageUrl'] ?? '') as String,
        galleryUrls:
            (json['galleryUrls'] as List?)?.map((e) => e as String).toList() ??
                const [],
        departments:
            (json['departments'] as List?)?.map((e) => e as String).toList() ??
                const [],
        facilities: (json['facilities'] as List?)
                ?.map((e) => HospitalFacility.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        about: (json['about'] ?? '') as String,
        openHours: (json['openHours'] ?? '') as String,
        distanceKm: ((json['distanceKm'] ?? 0) as num).toDouble(),
        latitude: ((json['latitude'] ?? 0) as num).toDouble(),
        longitude: ((json['longitude'] ?? 0) as num).toDouble(),
      );
}

/// A facility/amenity offered at a hospital, with an icon label.
class HospitalFacility {
  final String label;
  final String icon; // material icon code point key, resolved in widget

  const HospitalFacility(this.label, this.icon);

  factory HospitalFacility.fromJson(Map<String, dynamic> json) =>
      HospitalFacility(
        (json['label'] ?? '') as String,
        (json['icon'] ?? '') as String,
      );
}

File: backend/src/models/Hospital.js
'use strict';

const mongoose = require('mongoose');

/**
 * Hospital / clinic. This is the FULL shape the patient app's `Hospital` model
 * expects (gallery, facilities, rating, map coords), which is a superset of the
 * hospital backend's trimmed `Hospital`. Storing the superset keeps both
 * services happy on a shared `hospitals` collection — the hospital backend
 * simply ignores the extra fields.
 *
 * `distanceKm` is a demo/display value (there's no real geolocation here); a
 * production app would compute it from the device's location against
 * latitude/longitude.
 */
const facilitySchema = new mongoose.Schema(
  {
    label: { type: String, required: true }, // e.g. "24/7 Emergency"
    icon: { type: String, required: true }, // icon key resolved client-side
  },
  { _id: false }
);

const hospitalSchema = new mongoose.Schema(
  {
    _id: { type: String }, // "h1".."h5"
    name: { type: String, required: true },
    address: { type: String, default: '' },
    city: { type: String, default: '' },
    rating: { type: Number, default: 0, min: 0, max: 5 },
    phone: { type: String, default: '' },
    imageUrl: { type: String, default: '' },
    galleryUrls: { type: [String], default: [] },
    departments: { type: [String], default: [] },
    facilities: { type: [facilitySchema], default: [] },
    about: { type: String, default: '' },
    openHours: { type: String, default: '' }, // e.g. "Open 24 hours"
    distanceKm: { type: Number, default: 0 },
    latitude: { type: Number, default: 0 },
    longitude: { type: Number, default: 0 },
  },
  { _id: false, timestamps: true }
);

module.exports = mongoose.model('Hospital', hospitalSchema);
